neurots/generate/orientations.py

```diff
@@ -360,3 +438,216 @@ def compute_interval_n_tree(soma, n_trees, rng=np.random):
+            def _prob(angle):
+                return 0.5 * (
+                    expit((angle - scale_low) / rate_low) + expit((-angle - scale_high) / rate_high)
+                )
+
+            p = _prob(angle) / max(_prob(np.linspace(0, np.pi, 100)))
+            if with_density:
+                p *= np.sin(angle)
+            return np.clip(p, 0.0, 1.0)
+
+        return double_prob
+
+    raise ValueError(
+        f"The '{form}' value is unknown, it should be one of ['flat', 'step', 'double_step']"
+    )
+
+
+def _fit_single_3d_angles(data, neurite_type, morph_class, fit_params=None):
+    """Fit function to distribution of 3d angles for a neurite_type.
+
+    Args:
+        data (dict): bins and weights data from input_distribution
+        neurite_type (str): neurite_type to consider
+        morph_class (str): morph_class of the neuron (with_apical or without_apical)
+        fit_params (dict): specific fit parameters such as form and bounds to overwrite the defaults
+    """
+    _fit_params = deepcopy(fit_3d_angles_params)
+    if fit_params is not None:
+        _fit_params[morph_class][neurite_type].update(fit_params)
+
+    form = _fit_params[morph_class][neurite_type]["form"]
+    if form != "flat":
+        function = get_probability_function(form, with_density=True)
+
+        try:
+            popt = curve_fit(
+                function,
+                data["bins"],
+                data["weights"],
+                bounds=_fit_params[morph_class][neurite_type]["bounds"],
+            )[0]
+        except RuntimeError:
+            warnings.warn("Cannot fit some trunk angles, we fallback to flat distribution")
+            form = "flat"
+            popt = np.array([])
+    else:
+        popt = np.array([])
+    return {"form": form, "params": popt.tolist()}
+
+
+def _get_fit_params_from_input_parameters(parameters):
+    """Get parameter dict for fits from tmd_parameters."""
+    if "values" in parameters["orientation"] and parameters["orientation"]["values"] is not None:
+        form = parameters["orientation"]["values"].get("form")
+        bounds = parameters["orientation"]["values"].get("bounds")
+        if form is not None:
+            if bounds is None:
+                bounds = fit_3d_angles_bounds[form]
+            if parameters["orientation"]["values"].get("params") is None:
+                return {"form": form, "bounds": bounds}
+    return None
+
+
+def check_3d_angles(tmd_parameters):
+    with_3d = False
+    for neurite_type in tmd_parameters["grow_types"]:
+
+        if (
+            tmd_parameters[neurite_type]["orientation"] is None
+            or "mode" not in tmd_parameters[neurite_type]["orientation"]
+        ):
+            if with_3d:
+                raise Exception("Only partial 3d_angle parameters are present")
+        elif tmd_parameters[neurite_type]["orientation"]["mode"] in _3d_angles_mapping:
+            with_3d = True
+    return with_3d
```

Comment:
```suggestion
    with_3d = False
    count_with_3d = 0
    for neurite_type in tmd_parameters["grow_types"]:
        if tmd_parameters[neurite_type]["orientation"].get("mode") in _3d_angles_mapping:
            with_3d = True
            count_with_3d += 1
    if with_3d and count_with_3d != len(tmd_parameters["grow_types"]):
        raise Exception("Only partial 3d_angle parameters are present")
    return with_3d
```
As far as I can see, with your implementation the exception is not raised if only the last `neurite_type` uses 3d angles.. I think it's not what you want, so this change will raise if at least one `neurite_type` is 3d but not all of them.

Reply:
yes, thanks!